flocker/node/agents/test/test_blockdevice.py

```diff
@@ -1198,16 +1205,231 @@ class DidNotConverge(Exception):
     """
 
 
-class BlockDeviceCalculatorTests(SynchronousTestCase):
+class _WriteVerifyingExternalClient(object):
     """
-    Tests for ``BlockDeviceCalculator``.
+    Representation of an external client of the Flocker API. This client can be
+    used to try to write to paths on the system and verify that the writes are
+    persisted to the underlying blockdevice.
+
+    :ivar _known_mountpoints: Known mountpoints of the dataset. Once flocker
+        has returned a path to an external agent, there is no promising how
+        long that agent might cache that information before it tries to write
+        to that location. This agent keeps track of all mountpoints it has ever
+        seen and tries to write to all of them.
+    :ivar _device_path: The most up-to-date device path of the blockdevice that
+        backs the dataset. This is a backdoor best-effort way to verify that
+        flocker has persisted writes to the underlying blockdevice. Stored in
+        an attribute so that the test harness can inspect the blockdevice even
+        after the blockdevice deployer stops reporting the blockdevice in the
+        ``NodeState``.
+    :ivar _dataset_id: The dataset_id of the dataset under test.
+    :ivar _node_uuid: The node_uuid of the node that is being tested.
+    :ivar _testing_mountroot: A mountroot to construct moutpoints under.
+    :ivar _blockdevice_manager: An ``IBlockDeviceManager`` provider that can be
+        used to mount block devices.
     """
-    def setUp(self):
-        self.deployer = create_blockdevicedeployer(self)
 
-    def teardown_example(self, token):
+    def __init__(self, dataset_id, node_uuid, mountroot, blockdevice_manager):
+        self._known_mountpoints = set()
+        self._device_path = None
+        self._dataset_id = dataset_id
+        self._node_uuid = node_uuid
+        self._testing_mountroot = mountroot
+        self._blockdevice_manager = blockdevice_manager
+
+    def _has_file_with_content(self, filename, content):
+        """
+        Returns whether the blockdevice at self._device_path has a file with
+        the given name.
+
+        :param unicode filename: The name of the file to check for.
+        :param unicode content: The expected content of the file.
+
+        :returns: Whether the blockdevice that backs ``self._dataset_id`` has a
+            file with the given name that contains the given content.
+        """
+        test_mountpoint = self._testing_mountroot.child(str(uuid4()))
+        test_mountpoint.makedirs()
+
+        # If the device is already mounted, read the file from the existing
+        # mountpoint.
```

Comment:
As discussed on slack, I think it is fine to remount an already mounted block device, and am concerned about relying on interacting with more of the system under test than is necessary. Please remove this check and just always use the fallback to mounting the device on a fresh mount.